src/misc/umap.c

```diff
@@ -0,0 +1,846 @@
+                    MATRIX(jTj, j1, j2) += MATRIX(jacobian, i, j1) * MATRIX(jacobian, i, j2);
+                }
+                MATRIX(jTr, j1, 0) += MATRIX(jacobian, i, j1) * VECTOR(residuals)[i];
+            }
+        }
+        /* LAPACK puts solution into jTr */
+        IGRAPH_CHECK(igraph_lapack_dgesv(&jTj, 0, &jTr, &lapack_info));
+
+        /* This might go wrong, in which case we should fail graciously */
+        if (lapack_info != 0) {
+            igraph_vector_destroy(&x);
+            igraph_vector_destroy(&residuals);
+            igraph_matrix_destroy(&jacobian);
+            igraph_matrix_destroy(&jTj);
+            igraph_matrix_destroy(&jTr);
+            igraph_vector_destroy(&powb);
+            IGRAPH_FINALLY_CLEAN(6);
+            IGRAPH_ERROR("Singular matrix in the estimation of a and b for UMAP",  IGRAPH_EINVAL);
+        }
+
+        da = -MATRIX(jTr, 0, 0);
+        db = -MATRIX(jTr, 1, 0);
+
+        /* Improvement over GN: rough exponential line search for best delta
+         * start from largest change, and keep shrinking as long as we are going down
+         * */
+        squared_sum_res_old = squared_sum_res;
+        IGRAPH_CHECK(igraph_i_umap_get_ab_residuals(&residuals, &squared_sum_res, nr_points, a + da,
+                    b + db, &powb, &x, min_dist));
+
+#ifdef UMAP_DEBUG
+        printf("start line search, SSR before delta: %f, current SSR:, %f\n", squared_sum_res_old,
+                squared_sum_res);
+#endif
+        for (int k = 0; k < 30; k++) {
+            /* Try new parameters */
+            da /= 2.0;
+            db /= 2.0;
+            squared_sum_res_tmp = squared_sum_res;
+            IGRAPH_CHECK(igraph_i_umap_get_ab_residuals(&residuals, &squared_sum_res, nr_points,
+                        a + da, b + db, &powb, &x, min_dist));
+
+            /* Compare and if we are going back uphill, undo last step and break */
+#ifdef UMAP_DEBUG
+            printf("during line search, k = %d, old SSR:, %f, new SSR (half a,b):, %f\n", k,
+                    squared_sum_res_tmp, squared_sum_res);
+#endif
+            if (squared_sum_res > squared_sum_res_tmp - tol) {
+                da *= 2;
+                db *= 2;
+                break;
+            }
+        }
+#ifdef UMAP_DEBUG
+        printf("end of line search and iteration, squared_sum_res: %f \n\n", squared_sum_res_tmp);
+#endif
+
+        /* assign a, b*/
+        a += da;
+        b += db;
+
+    }
+
+    /* Free memory and tidy up stack */
+    igraph_vector_destroy(&x);
+    igraph_vector_destroy(&residuals);
+    igraph_matrix_destroy(&jacobian);
+    igraph_matrix_destroy(&jTj);
+    igraph_matrix_destroy(&jTr);
+    igraph_vector_destroy(&powb);
+    IGRAPH_FINALLY_CLEAN(6);
+
+#ifdef UMAP_DEBUG
+    printf("a, b: %f %f\n", a, b);
+#endif
+
+    *a_p = a;
+    *b_p = b;
+
+    return (IGRAPH_SUCCESS);
+
+}
+
+/* cross-entropy */
+static igraph_error_t igraph_compute_cross_entropy(const igraph_t *graph,
```

Comment:
If this function is used only for debugging, it should also be conditional on `UMAP_DEBUG`.